sequencer/src/state_signature/relay_server.rs

```diff
@@ -0,0 +1,259 @@
+    /// Get the latest available signatures package.
+    /// # Errors
+    /// Errors if there's no available signatures package.
+    fn get_latest_signature_package(&self) -> Result<StateSignaturesPackage, Error>;
+
+    /// Post a signature to the relay server
+    /// # Errors
+    /// Errors if the signature is invalid, already posted, or no longer needed.
+    fn post_signature(
+        &mut self,
+        key: StateVerKey,
+        state: LightClientState,
+        signature: StateSignature,
+    ) -> Result<(), Error>;
+}
+
+impl StateRelayServerDataSource for StateRelayServerState {
+    fn get_latest_signature_package(&self) -> Result<StateSignaturesPackage, Error> {
+        match &self.latest_available_package {
+            Some(package) => Ok(package.clone()),
+            None => Err(tide_disco::error::ServerError::catch_all(
+                StatusCode::NotFound,
+                "The light client state signatures are not ready.".to_owned(),
+            )),
+        }
+    }
+
+    fn post_signature(
+        &mut self,
+        key: StateVerKey,
+        state: LightClientState,
+        signature: StateSignature,
+    ) -> Result<(), Error> {
+        if (state.block_height as u64) <= self.latest_block_height.unwrap_or(0) {
+            return Err(tide_disco::error::ServerError::catch_all(
+                StatusCode::Locked,
+                "The posted signature is no longer needed.".to_owned(),
+            ));
+        }
+        let one = U256::one();
+        let weight = self.known_nodes.get(&key).unwrap_or(&one);
+        // TODO(Chengyu): We don't know where to fetch the stake table yet.
+        // Related issue: [https://github.com/EspressoSystems/espresso-sequencer/issues/1022]
+        // .ok_or(tide_disco::error::ServerError::catch_all(
+        //     StatusCode::Unauthorized,
+        //     "The posted key is not found in the stake table.".to_owned(),
+        // ))?;
+        let state_msg: [FieldType; 7] = (&state).into();
+        if StateSignatureScheme::verify(&(), &key, state_msg, &signature).is_err() {
+            return Err(tide_disco::error::ServerError::catch_all(
+                StatusCode::BadRequest,
+                "The posted signature is not valid.".to_owned(),
+            ));
+        }
+        let block_height = state.block_height as u64;
+        // TODO(Chengyu): this serialization should be removed once `LightClientState` implements `Eq`.
+        let bytes = to_bytes!(&state).unwrap();
+        let package = self.packages.entry(block_height).or_insert_with(|| {
+            self.queue.insert(block_height);
+            StateSignaturesPackage {
+                state,
+                signatures: Default::default(),
+                accumulated_weight: U256::from(0),
+            }
+        });
+        let incoming_bytes = to_bytes!(&package.state).unwrap();
+        if bytes != incoming_bytes {
+            return Err(tide_disco::error::ServerError::catch_all(
```

Comment:
Is there an attack here where if the first node to post a signature signed the wrong state, all subsequent signatures on the correct state will be rejected? I think to prevent this we would have to keep track of all states we have seen for a given block height until one of them has a quorum of signatures (or at least f+1 signatures). It's a bit complicated but maybe necessary?

Reply:
Yes you are right. I was a bit lazy before and ignored this attack.

Reply:
Fix this by adding another HashMap: [1dde57c](https://github.com/EspressoSystems/espresso-sequencer/pull/1016/commits/1dde57c61e4683b1fde67bb5bde9e666a563311b)